cli/deno_dir.rs

```diff
@@ -1,145 +1,30 @@
-  pub source_code: Vec<u8>,
-}
-
-impl SourceFile {
-  // TODO(bartlomieju): this method should be implemented on new `CompiledSourceFile`
-  // trait and should be handled by "compiler pipeline"
-  pub fn js_source(&self) -> String {
-    if self.media_type == msg::MediaType::TypeScript {
-      panic!("TypeScript module has no JS source, did you forget to run it through compiler?");
-    }
-
-    // TODO: this should be done by compiler and JS module should be returned
-    if self.media_type == msg::MediaType::Json {
-      return format!(
-        "export default {};",
-        str::from_utf8(&self.source_code).unwrap()
-      );
-    }
-
-    // it's either JS or Unknown media type
-    str::from_utf8(&self.source_code).unwrap().to_string()
-  }
-}
-
-pub type SourceFileFuture =
-  dyn Future<Item = SourceFile, Error = ErrBox> + Send;
-
-/// This trait implements synchronous and asynchronous methods
-/// for file fetching.
-///
-/// Implementors might implement caching mechanism to
-/// minimize number of fs ops/net fetches.
-pub trait SourceFileFetcher {
-  fn check_if_supported_scheme(url: &Url) -> Result<(), ErrBox>;
-
-  fn fetch_source_file_async(
-    self: &Self,
-    specifier: &ModuleSpecifier,
-  ) -> Box<SourceFileFuture>;
-
-  /// Required for TS compiler.
-  fn fetch_source_file(
-    self: &Self,
-    specifier: &ModuleSpecifier,
-  ) -> Result<SourceFile, ErrBox> {
-    tokio_util::block_on(self.fetch_source_file_async(specifier))
-  }
-}
-
-// TODO: this list should be implemented on SourceFileFetcher trait
-const SUPPORTED_URL_SCHEMES: [&str; 3] = ["http", "https", "file"];
-
-/// Simple struct implementing in-process caching to prevent multiple
-/// fs reads/net fetches for same file.
-#[derive(Clone, Default)]
-pub struct SourceFileCache(Arc<Mutex<HashMap<String, SourceFile>>>);
-
-impl SourceFileCache {
-  pub fn set(&self, key: String, source_file: SourceFile) {
-    let mut c = self.0.lock().unwrap();
-    c.insert(key, source_file);
-  }
-
-  pub fn get(&self, key: String) -> Option<SourceFile> {
-    let c = self.0.lock().unwrap();
-    match c.get(&key) {
-      Some(source_file) => Some(source_file.clone()),
-      None => None,
-    }
-  }
-}
 
 /// `DenoDir` serves as coordinator for multiple `DiskCache`s containing them
 /// in single directory that can be controlled with `$DENO_DIR` env variable.
 #[derive(Clone)]
-// TODO(bartlomieju): try to remove `pub` from fields
 pub struct DenoDir {
   // Example: /Users/rld/.deno/
   pub root: PathBuf,
-  // This is where we cache compilation outputs. Example:
-  // /Users/rld/.deno/gen/http/github.com/ry/blah.js
-  // TODO: this cache can be created using public API by TS compiler
-  pub gen_cache: DiskCache,
-  // /Users/rld/.deno/deps/http/github.com/ry/blah.ts
-  pub deps_cache: DiskCache,
-
-  pub progress: Progress,
-
-  source_file_cache: SourceFileCache,
-
-  use_disk_cache: bool,
-  no_remote_fetch: bool,
+  cache_map: Arc<Mutex<HashMap<String, DiskCache>>>,
```

Comment:
Can you give a triple slash comment here describing the typical keys here? I guess it's "deps" and "gen" ?

Reply:
> Can you add a comment here describing the typical keys here? I guess it's "deps" and "gen" ?

Sure

> Will it ever be anything else other than "deps" and "gen"? If not, what's the benefit of having this generalized HashMap structure?

It will if we introduce other compilers - then each compiler registers cache for itself.
`import { foo } from "./foo.rs` is still on my roadmap and I should be able to implement it once #2686 lands.

Reply:
I see, so 'gen' and 'deps' belong to the typescript compiler, but there may be others in the future...

What about using an enum instead of String for the key?

Reply:
Actually:
-  `gen` belongs to `TsCompiler` and is used and store compiler outputs and source maps
- `deps` belongs to `FileFetcher` and is used to store remote modules locally

Yeah it definitely makes sense to use an enum. Will update :+1:

Reply:
I'm gonna bring back `DenoDir.deps_cache` and `DenoDir.gen_cache` back. Keeping them in hash map is an overkill now...